### PR TITLE
Add --parallel flag to the 'pull' command

### PR DESCRIPTION
**Pull request by maallen**
Adds a `--parallel` flag to the pull command which runs the generation of localized assets in parallel across multiple (if available) worker nodes via Quartz jobs.

### CODE REVIEW

File: webapp/src/main/java/com/box/l10n/mojito/rest/asset/LocaleInfo.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.box.l10n.mojito.rest.asset;
+
+public class LocaleInfo {
+
+  Long localeId;
+
+  String bcp47Tag;
```
Comment (garionpin) [inline webapp/src/main/java/com/box/l10n/mojito/rest/asset/LocaleInfo.java:7]: Same question as before around needing both vs. just one of them.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 2]
```diff
@@ -158,6 +161,12 @@ public class PullCommand extends Command {
               + "A file will be generated with the pull run name.")
   Boolean recordPullRun = false;
 
+  @Parameter(
+      names = {"--parallel"},
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:165]: I can add to the `--parallel` parameter description that it uses a async web service and that when `--parallel` is specified with `--async-ws` then `--async-ws` is ignored.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:1]: My concern with doing an all at once approach is that we send the source content across the wire to the backend to be translated, in the case of Android the request size would be quite large if we were to send an entire repo's worth of string.xml content across in a single async request. 

It might be possible for us to send all the individual requests first to the server, store the returned pollable task ids and then start polling for them from the client once we've queued up all the requests.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 3]
```diff
@@ -496,6 +594,45 @@ LocalizedAssetBody getLocalizedAssetBodySync(
...
+            assetByPathAndRepositoryId.getId(),
+            assetContent,
+            repositoryLocales,
+            localeIdToOutputTagsMap,
+            sourceFileMatch.getFileType().getFilterConfigIdOverride(),
+            filterOptions,
+            status,
+            inheritanceMode,
+            pullRunName);
+    commandHelper.waitForPollableTask(localizedAssetForContentParallel.getId());
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:619]: waitForPollableTask can take a WaitForPollableTaskListener. if there enough information in the pollabletask, it could be used to display an improved status update in the console

File: webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    for (LocaleInfo localeInfo : input.getLocaleInfos()) {
+      String outputTag =
+          localeInfo.getOutputBcp47tag() != null
+              ? localeInfo.getOutputBcp47tag()
+              : localeInfo.getBcp47Tag();
+      QuartzJobInfo<LocalizedAssetBody, LocalizedAssetBody> quartzJobInfo =
+          QuartzJobInfo.newBuilder(GenerateLocalizedAssetJob.class)
+              .withInlineInput(false)
+              .withParentId(getParentId())
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java:27]: Can we, thanks to this, show a status in the CLI console of all tasks while polling on the main task?

File: webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+              .withInput(createLocalizedAssetBody(localeInfo, input))
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java:28]: thinking out loud... So we duplicate the source for each job, vs having the source saved once and read by multiple jobs? I guess it duplicates the data but the jobs are fully isolated. Any thoughts on that in general?

File: webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    return input;
+  }
+
+  protected long getParentId() {
+    return getCurrentPollableTask().getId();
+  }
+
+  private LocalizedAssetBody createLocalizedAssetBody(
+      LocaleInfo localeInfo, MultiLocalizedAssetBody input) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java:45]: Maybe rename "input" to something that remind the type name

File: webapp/src/main/java/com/box/l10n/mojito/rest/asset/MultiLocalizedAssetBody.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  public void addGenerateLocalizedAddedJobIdToMap(
+      String bcp47Tag, Long generateLocalizedAssetJobId) {
+    this.generateLocalizedAssetJobIds.put(bcp47Tag, generateLocalizedAssetJobId);
+  }
+
+  public String getSchedulerName() {
+    return schedulerName;
+  }
+
+  public void setSchedulerName(String schedulerName) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/rest/asset/MultiLocalizedAssetBody.java:124]: I don't see this being set anywhere but the test, is it?

File: webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java (hunk 5 above)
Comment (maallen) [reply webapp/src/main/java/com/box/l10n/mojito/service/tm/GenerateMultiLocalizedAssetJob.java:28]: Doing it this way using isolated jobs means we can re-use the existing single locale generation logic without updates meaning we avoid introducing regressions into the existing flows. 

Using this approach we only send the source from the client to the server once where previously we'd send it with every individual locale, I felt re-using the existing generation locale asset logic was a better fit for 2 reasons:

1. We know it works 😄 
2. Easier maintenance going forward as pull, async-ws & parallel all use the same code for individual locale asset generation.

File: webapp/src/main/java/com/box/l10n/mojito/rest/asset/MultiLocalizedAssetBody.java (hunk 7 above)
Comment (maallen) [reply webapp/src/main/java/com/box/l10n/mojito/rest/asset/MultiLocalizedAssetBody.java:124]: It's used when creating the individual `GenerateLoclizedAssetJobs` quartzJobInfos, it ensures they get scheduled to the same scheduler as the parent

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 8]
```diff
@@ -383,7 +467,21 @@ void writeLocalizedAssetToTargetDirectory(
 
     Path relativeTargetFilePath = commandDirectories.relativizeWithUserDirectory(targetPath);
 
-    consoleWriter.a(" --> ").fg(Color.MAGENTA).a(relativeTargetFilePath.toString()).println();
+    if (!isParallel) {
+      consoleWriter.a(" --> ").fg(Color.MAGENTA).a(relativeTargetFilePath.toString()).println();
+    } else {
+      synchronized (this) {
+        consoleWriter
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:474]: Console output has changed with the implementation change to push all requests to the server first and then poll for the generated files in parallel.

File: webapp/src/main/java/com/box/l10n/mojito/rest/asset/AssetWS.java [hunk 9]
```diff
@@ -243,6 +244,42 @@ public PollableTask getLocalizedAssetForContentAsync(
...
+      multiLocalizedAssetBody.setAssetId(assetId);
+    }
+
+    multiLocalizedAssetBody.setSchedulerName(schedulerName);
+
+    Asset asset = assetRepository.getOne(assetId);
+    meterRegistry
+        .counter(
+            "assetWS.getLocalizedAssetForContentParallel",
+            Tags.of("repositoryId", asset.getRepository().getId().toString()))
```
Comment (garionpin) [inline webapp/src/main/java/com/box/l10n/mojito/rest/asset/AssetWS.java:269]: For any new metrics - could we add the repository name in as well?
Having just the ID makes it very hard to parse/interpret the data when using the metrics.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 10]
```diff
@@ -213,24 +233,86 @@ public void execute() throws CommandException {
...
+    }
+  }
+
+  private void pollForLocalizedFiles() {
+    pollableTaskIdToFileMatchMap
+        .entrySet()
+        .parallelStream()
+        .forEach(
+            entry -> {
+              commandHelper.waitForPollableTaskSilencedOutput(entry.getKey());
```
Comment (garionpin) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:299]: Would there be any concerns with having many parallel tasks that (eventually) call waitForPollableTask which, essentially, 99% of their time, do Thread.Sleep()? 

I'm assuming that for our volumes there isn't a practical concern that would warrant doing this with a Executors.newScheduledThreadPool().scheduleAtFixedRate(..) pattern instead to avoid needing to always have the Threads allocated - wanted to check just in case.

**Review by garionpin — APPROVED** [review]

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 10 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:299]: In practice the `parallelStream` uses numOfCPUs - 1 threads so we'll only ever have that many active tasks running on the client at once. 

As the task itself is quite small (wait for server side task to finish, then write all localized assets to file) I don't think we require the extra overhead of using an Executor thread pool approach here.

Comment (aurambaj) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:299]: Looking from far away... this might not be optimal polling. Say the threads get stuck on the tasks that finish last, then the rest that could be download in the meantime is not and it is delaying the process.

It feel like we should have a queue of tasks to process, but if the task is not finished we move it to the end of the queue and start checking others. Multi-threading also seem good to have so I'd be looking into ExecutorService/ThreadPoolExecutors maybe? Thoughts?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 11]
```diff
@@ -282,16 +439,38 @@ void generateLocalizedFilesWithLocaleMaping(
...
-      generateLocalizedFile(
-          repository, sourceFileMatch, filterOptions, outputBcp47tag, repositoryLocale);
+    List<RepositoryLocale> repositoryLocales =
+        localeMappings.entrySet().stream()
+            .map(entry -> getRepositoryLocaleForOutputBcp47Tag(entry.getKey()))
+            .distinct()
+            .collect(Collectors.toList());
+
+    if (isParallel) {
+      generateLocalizedFiles(repository, sourceFileMatch, filterOptions, repositoryLocales);
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:451]: should this mention "parallel" in the method name?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 12]
```diff
@@ -282,16 +439,38 @@ void generateLocalizedFilesWithLocaleMaping(
...
+    } else {
+      for (Map.Entry<String, String> localeMapping : localeMappings.entrySet()) {
+        String outputBcp47tag = localeMapping.getKey();
+        RepositoryLocale repositoryLocale = getRepositoryLocaleForOutputBcp47Tag(outputBcp47tag);
+        generateLocalizedFile(
+            repository, sourceFileMatch, filterOptions, outputBcp47tag, repositoryLocale);
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:457]: and this be in a method?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 13]
```diff
@@ -264,9 +359,71 @@ void generateLocalizedFilesWithoutLocaleMapping(
 
     logger.debug("Generate localized files (without locale mapping)");
 
-    for (RepositoryLocale repositoryLocale : repositoryLocalesWithoutRootLocale.values()) {
-      generateLocalizedFile(repository, sourceFileMatch, filterOptions, null, repositoryLocale);
+    if (isParallel) {
+      generateLocalizedFiles(
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:363]: should this mention "parallel" in the method name?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 14]
```diff
@@ -264,9 +359,71 @@ void generateLocalizedFilesWithoutLocaleMapping(
...
+          repository,
+          sourceFileMatch,
+          filterOptions,
+          Lists.newArrayList(repositoryLocalesWithoutRootLocale.values()));
+    } else {
+      for (RepositoryLocale repositoryLocale : repositoryLocalesWithoutRootLocale.values()) {
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:369]: and this be in a method?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java [hunk 15]
```diff
@@ -194,6 +206,10 @@ public void execute() throws CommandException {
 
     repository = commandHelper.findRepositoryByName(repositoryParam);
 
+    if (isParallel) {
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:209]: Just wondering... at this point, is there really a value in keeping the 2 implementation mixed, vs having to distinct path as it is seem there are quite a few "isParallel"? I can image that we'll be removing the non-parallel eventually?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/CommandHelper.java [hunk 16]
```diff
@@ -286,6 +286,16 @@ public void waitForPollableTask(Long pollableId) throws CommandException {
     }
   }
 
+  public void waitForPollableTaskSilencedOutput(Long pollableId) throws CommandException {
+
+    try {
+      pollableTaskClient.waitForPollableTask(
+          pollableId, PollableTaskClient.NO_TIMEOUT, new CommandWaitForPollableTaskListener(true));
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/CommandHelper.java:293]: Wasn't aware of the null support, will use that instead.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 15 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:209]: I don't believe the should be separate cli commands as there's quite a bit of crossover between pull run recordings, locale mapping etc but I can look into moving the parallel specific implementation into a separate class which is called from the pull command execute method, then we'd just require a single if statement in the PullCommand to call this new class

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 10 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java:299]: I guess my view is how big of an impact is that situation likely to have? As the waiting threads 'active' task (write files to disk) is quite small, and that we have multiple threads waiting, I don't expect much of an impact realistically on performance. 

If we have one thread polling the longest tasks I accept it could be writing while it's waiting but when the task does complete it'll iterate through the other completed tasks immediately. In the queued scenario there's a possibility that our polling just misses the top of queue's finish, moves that job to the back of the queue, same then happens for the next job etc. I'd be surprised if we're talking about more than a few seconds performance gain either way.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 11 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java]: I've moved this method to a `PullCommandParallel` class now so I don't think the rename is still required

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 15 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java]: Pushed an update that moves the parallel logic to a separate class that is called from the PullCommand if `--parallel` flag is passed

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java (hunk 10 above)
Comment (aurambaj) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java]: > In practice the parallelStream uses numOfCPUs - 1 threads

Any thoughts on if we should potentially have more?

Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java]: As the majority of tasks done by the cli thread is i/o rather than cpu bound we probably could have more, my only minor concern would be contention around the lock on console write for writing to the console. However the recommendation for parallelStreams is to use the common pool rather than a custom pool, and the common pool is only increased via a global parameter which is why I left it as it was

**Review by aurambaj — APPROVED** [review]
